### DISCUSSION TITLE
How to merge points from an unstructured grid?

### DISCUSSION THREAD

**Discussion by rodrigomologni** [pyvista/pyvista#3490] [Q&A]
I would like to merge points from an unstructured grid, but `DataSetFilters.merge` is not reducing the number of points from 16 to 12. Why is not it working?

```py
import pyvista as pv
import numpy as np
import pyvista._vtk as vtk

cells = np.array([8, 0, 1, 2, 3, 4, 5, 6, 7, 8, 8, 9, 10, 11, 12, 13, 14, 15])
cell_type = np.array([vtk.VTK_HEXAHEDRON, vtk.VTK_HEXAHEDRON])
cell1 = np.array([[0, 0, 0],
                  [1, 0, 0],
                  [1, 1, 0],
                  [0, 1, 0],
                  [0, 0, 1],
                  [1, 0, 1],
                  [1, 1, 1],
                  [0, 1, 1]])
cell2 = np.array([[0, 0, 1],
                  [1, 0, 1],
                  [1, 1, 1],
                  [0, 1, 1],
                  [0, 0, 2],
                  [1, 0, 2],
                  [1, 1, 2],
                  [0, 1, 2]])
points = np.vstack((cell1, cell2)).astype(float)
grid = pv.UnstructuredGrid(cells, cell_type, points)
print(grid)
grid = grid.merge()
print(grid)
grid.plot(style="wireframe")
```
Output:
```sh
UnstructuredGrid (0x189d5f61640)
  N Cells:  2
  N Points: 16
  X Bounds: 0.000e+00, 1.000e+00
  Y Bounds: 0.000e+00, 1.000e+00
  Z Bounds: 0.000e+00, 3.000e+00
  N Arrays: 0

UnstructuredGrid (0x189d5f79820)
  N Cells:  2
  N Points: 16
  X Bounds: 0.000e+00, 1.000e+00
  Y Bounds: 0.000e+00, 1.000e+00
  Z Bounds: 0.000e+00, 3.000e+00
  N Arrays: 0
```
![image](https://user-images.githubusercontent.com/1608652/196444367-9110383d-8e34-4874-a7a2-69b1155ef4d0.png)

**Comment by banesullivan** [comment]
The `clean` filter is designed for this while the `merge` filter will not combine duplicate points. Unfortunately, the `clean` filter is only for `PolyData` at this time. Ref

- https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.clean.html?highlight=clean#pyvista.PolyDataFilters.clean
- https://vtk.org/doc/nightly/html/classvtkCleanPolyData.html#details

If you build VTK with VTKm, you could use the `vtkmCleanGrid` filter: https://vtk.org/doc/nightly/html/classvtkmCleanGrid.html#details

**Reply by rodrigomologni** [reply]
Hi @banesullivan!

Thank you for answer me!

I didn't know the `vtkmCleanGrid` class. I will try to use it. Thank you!

In `UnstructuredGrid.merge` documentation it says: "Can be used to merge points of adjacent cells when no grids are input".

The `merge` filter has the `merge_points` parameter and it works when the input dataset is an explicit structured grid. Why?

```py
grid = pv.UnstructuredGrid(cells, cell_type, points)
grid["BLOCK_I"] = [0, 0]
grid["BLOCK_J"] = [0, 0]
grid["BLOCK_K"] = [0, 1]
grid = grid.cast_to_explicit_structured_grid()
print(grid)
grid = grid.merge()
print(grid)
```
```
ExplicitStructuredGrid (0x1b626988be0)
  N Cells:  2
  N Points: 16
  X Bounds: 0.000e+00, 1.000e+00
  Y Bounds: 0.000e+00, 1.000e+00
  Z Bounds: 0.000e+00, 2.000e+00
  N Arrays: 3

UnstructuredGrid (0x1b626988b80)
  N Cells:  2
  N Points: 12
  X Bounds: 0.000e+00, 1.000e+00
  Y Bounds: 0.000e+00, 1.000e+00
  Z Bounds: 0.000e+00, 2.000e+00
  N Arrays: 3
```

**Reply by banesullivan** [reply]
Huh, this is different than what I thought/remembered. I'll try to dig into this...
